samples/Common.c

```diff
@@ -706,27 +716,62 @@ VOID sampleBandwidthEstimationHandler(UINT64 customData, DOUBLE maximumBitrate)
     DLOGV("received bitrate suggestion: %f", maximumBitrate);
 }
 
+// Sample callback for TWCC. Average packet is calculated with EMA. If average packet lost is <= 5%,
+// the current bitrate is increased by 5%. If more than 5%, the current bitrate
+// is reduced by percent lost. Bitrate update is allowed every second and is increased/decreased upto the limits
 VOID sampleSenderBandwidthEstimationHandler(UINT64 customData, UINT32 txBytes, UINT32 rxBytes, UINT32 txPacketsCnt, UINT32 rxPacketsCnt,
                                             UINT64 duration)
 {
-    UNUSED_PARAM(customData);
     UNUSED_PARAM(duration);
-    UNUSED_PARAM(rxBytes);
-    UNUSED_PARAM(txBytes);
+    UINT64 videoBitrate, audioBitrate;
+    UINT64 currentTimeMs, timeDiff;
     UINT32 lostPacketsCnt = txPacketsCnt - rxPacketsCnt;
-    UINT32 percentLost = lostPacketsCnt * 100 / txPacketsCnt;
-    UINT32 bitrate = 1024;
-    if (percentLost < 2) {
-        // increase encoder bitrate by 2 percent
-        bitrate *= 1.02f;
-    } else if (percentLost > 5) {
-        // decrease encoder bitrate by packet loss percent
-        bitrate *= (1.0f - percentLost / 100.0f);
-    }
-    // otherwise keep bitrate the same
-
-    DLOGV("received sender bitrate estimation: suggested bitrate %u sent: %u bytes %u packets received: %u bytes %u packets in %lu msec", bitrate,
-          txBytes, txPacketsCnt, rxBytes, rxPacketsCnt, duration / 10000ULL);
+    DOUBLE percentLost = (DOUBLE) ((txPacketsCnt > 0) ? (lostPacketsCnt * 100 / txPacketsCnt) : 0.0);
+    SampleStreamingSession* pSampleStreamingSession = (SampleStreamingSession*) customData;
+
+    if (pSampleStreamingSession == NULL) {
+        DLOGW("Invalid streaming session (NULL object)");
+        return;
+    }
+
+    // Calculate packet loss
+    pSampleStreamingSession->twccMetadata.averagePacketLoss =
+        EMA_ACCUMULATOR_GET_NEXT(pSampleStreamingSession->twccMetadata.averagePacketLoss, ((DOUBLE) percentLost));
+
+    currentTimeMs = GETTIME();
+    timeDiff = currentTimeMs - pSampleStreamingSession->twccMetadata.lastAdjustmentTimeMs;
+    if (timeDiff < TWCC_BITRATE_ADJUSTMENT_INTERVAL_SECONDS) {
```

Comment:
The unit here should be millis, timediff is in millis

Reply:
It essentially means we need a 1 second difference in millisecond format. I changed it to clarify, but this does not change the functionality in anyway. Still means 1 second.